sync/worker_pool.go

```diff
@@ -94,3 +81,17 @@ func (p *workerPool) GoWithTimeout(work Work, timeout time.Duration) bool {
 		return false
 	}
 }
+
+func (p *workerPool) Close() error {
+	if len(p.workCh) < cap(p.workCh) {
+		return errAllRoutinesNotFinished
+	}
+
+	set := atomic.CompareAndSwapInt64(&p.closed, 0, 1)
```

Comment:
I think there is a chance in between the check above and the atomic swap here that someone may call `p.Go`, so if you close the channel below before that call returns, the program will panic due to enqueuing to a closed channel.

Reply:
+1

I think we also need to change the semantics of the `select` statements if we add a `Close` method. Otherwise, once the channel is closed, `token := <-p.workCh` will return immediately, and we wouldn't know that it is because the channel is closed and not because a token was returned to the pool. So perhaps we should change the call to `token, ok := <-p.workCh` and check that `ok` is true?